src/System.IO.Pipelines/Gate.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// Simple awaitable gate - intended to synchronize a single producer with a single consumer to ensure the producer doesn't
+    /// produce until the consumer is ready. Similar to a <see cref="TaskCompletionSource{TResult}"/> but reusable so we don't have
+    /// to keep allocating new ones every time.
+    /// </summary>
+    /// <remarks>
+    /// The gate can be in one of two states: "Open", indicating that an await will immediately return and "Closed", meaning that an await
+    /// will block until the gate is opened. The gate is initially "Closed" and can be opened by a call to <see cref="Open"/>. Upon the completion
+    /// of an await, it will automatically return to the "Closed" state (this is done in the <see cref="GetResult"/> call that is injected by the
+    /// compiler's async/await logic).
+    /// </remarks>
+    internal class Gate : ICriticalNotifyCompletion
+    {
+        private static readonly Action _gateIsOpen = () => {};
+
+        private volatile Action _gateState;
+
+        /// <summary>
+        /// Returns a boolean indicating if the gate is "open"
+        /// </summary>
+        public bool IsCompleted => _gateState == _gateIsOpen;
```

Comment:
> I'd prefer if there was a reusable TCS in the framework, these 2 types are basically that.

I've considered that several times over the years.  I think it'd be reasonable to add something to TCS that would essentially allow it to dump its existing task and create a new one, but that's really not sufficient to replace these custom awaitables... for that, you'd not only need to make TCS reusable, but you'd need to allow a Task to go from completed to not completed, which could potentially break a whole lot of code if used incorrectly.  Such a task-resetting feature could only be used in a closed environment where the producer and consumer of the task coordinate, as otherwise the producer won't know when it's safe to reset (the same kind of issue that exists with pooled data structures in general).